src/core/sensor/qgssensormodel.cpp

```diff
@@ -0,0 +1,267 @@
+    email                : mathieu at opengis dot ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgssensormodel.h"
+
+#include "qgis.h"
+#include "qgssensormanager.h"
+#include "qgsabstractsensor.h"
+
+
+QgsSensorModel::QgsSensorModel( QgsSensorManager *manager, QObject *parent )
+  : QAbstractItemModel( parent )
+  , mSensorManager( manager )
+{
+  connect( mSensorManager, &QgsSensorManager::sensorAdded, this, &QgsSensorModel::sensorAdded );
+  connect( mSensorManager, &QgsSensorManager::sensorAboutToBeRemoved, this, &QgsSensorModel::sensorRemoved );
+
+  connect( mSensorManager, &QgsSensorManager::sensorNameChanged, this, &QgsSensorModel::sensorNameChanged );
+  connect( mSensorManager, &QgsSensorManager::sensorStatusChanged, this, &QgsSensorModel::sensorStatusChanged );
+  connect( mSensorManager, &QgsSensorManager::sensorDataCaptured, this, &QgsSensorModel::sensorDataCaptured );
+}
+
+QVariant QgsSensorModel::data( const QModelIndex &index, int role ) const
+{
+  if ( index.row() < 0 || index.row() >= rowCount( QModelIndex() ) )
+    return QVariant();
+
+  switch ( role )
+  {
+    case Qt::DisplayRole:
+    case Qt::ToolTipRole:
+    case Qt::EditRole:
+    {
+      switch ( index.column() )
+      {
+        case Name:
+        {
+          QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
+          return sensor ? sensor->name() : QString();
+        }
+
+        case LastValue:
+        {
+          QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
+          if ( sensor )
+          {
+            switch ( sensor->status() )
+            {
+              case Qgis::DeviceConnectionStatus::Disconnected:
+                return tr( "Disconnected" );
+
+              case Qgis::DeviceConnectionStatus::Connecting:
+                return tr( "Connecting" );
+
+              case Qgis::DeviceConnectionStatus::Connected:
+                return sensor->data().lastValue.toString();
+            }
+          }
+          return QVariant();
+        }
+
+        default:
+          break;
+      }
+
+      return QVariant();
+    }
+
+    case SensorType:
+    {
+      QgsAbstractSensor *sensor = mSensorManager->sensor( mSensorIds[index.row()] );
+      return sensor ? sensor->type() : QString();
+    }
+
+    case SensorId:
+    {
+      return mSensorIds[index.row()];
```

Comment:
potential crash here if index.row() isn't present in the map -- would be solved by the above comment though.